### PR TITLE
Unify transaction pushing

### PR DESCRIPTION
**Pull request by Quexington**
Recreation of #16346 @ main

Source hash: 3b1fcf2e026c689906e50df4b5eee054dc96a0bd
Remaining commits: 14

### CODE REVIEW

**Comment by emlowe** [comment]
this seems ok to me - probably could use a second look from Arvid or Amine

File: chia/pools/pool_wallet.py [hunk 1]
```diff
@@ -889,6 +889,7 @@ async def claim_pool_rewards(
             )
             assert fee_tx.spend_bundle is not None
             full_spend = SpendBundle.aggregate([fee_tx.spend_bundle, claim_spend])
+            fee_tx = dataclasses.replace(fee_tx, spend_bundle=None)
```
Comment (arvidn) [inline chia/pools/pool_wallet.py:892]: I don't understand why you're changing `fee_tx` here

File: chia/wallet/wallet_state_manager.py [hunk 2]
```diff
@@ -2214,20 +2213,43 @@ async def coin_added(
...
+        agg_spend: SpendBundle = SpendBundle.aggregate(
+            [tx.spend_bundle for tx in tx_records if tx.spend_bundle is not None]
+        )
+        actual_spend_involved: bool = agg_spend != SpendBundle([], G2Element())
+        if merge_spends and actual_spend_involved:
+            tx_records = [
+                dataclasses.replace(
+                    tx,
+                    spend_bundle=agg_spend if i == 0 else None,
+                    name=agg_spend.name() if i == 0 else bytes32.secret(),
```
Comment (arvidn) [inline chia/wallet/wallet_state_manager.py:2232]: `bytes32.secret()` is not deterministic. I take it that's not a problem, right?

File: tests/wallet/sync/test_wallet_sync.py [hunk 3]
```diff
@@ -1144,7 +1144,7 @@ async def test_dusted_wallet(
     )
     assert len(txs) == 1
     assert txs[0].spend_bundle is not None
-    await farm_wallet_node.wallet_state_manager.add_pending_transaction(txs[0])
```
Comment (arvidn) [inline tests/wallet/sync/test_wallet_sync.py:1147]: this appears to only add the *first* transaction. the new version seems to add all of them. Is this change deliberate?

File: tests/wallet/nft_wallet/test_nft_wallet.py [hunk 4]
```diff
@@ -160,7 +160,7 @@ async def test_nft_wallet_creation_automatically(
     )
     assert len(txs) == 1
     assert txs[0].spend_bundle is not None
-    await wallet_node_0.wallet_state_manager.add_pending_transaction(txs[0])
```
Comment (arvidn) [inline tests/wallet/nft_wallet/test_nft_wallet.py:163]: this seems to change behavior. adding only the first transaction vs. all of them

File: tests/wallet/did_wallet/test_did.py [hunk 5]
```diff
@@ -234,7 +234,7 @@ async def test_creation_from_backup_file(self, self_hostname, three_wallet_nodes
             test_message_spend_bundle,
         )
         assert txs[0].spend_bundle is not None
-        await did_wallet_2.wallet_state_manager.add_pending_transaction(txs[0])
```
Comment (arvidn) [inline tests/wallet/did_wallet/test_did.py:237]: is this change of adding the first transaction vs. all transactions deliberate?

File: tests/wallet/nft_wallet/test_nft_wallet.py [hunk 6]
```diff
@@ -307,7 +307,7 @@ async def ensure_wallet_sync() -> None:
...
```
Comment (arvidn) [inline tests/wallet/nft_wallet/test_nft_wallet.py:310]: this seems to be a recurring patter. I'll just point out all the places I find that look suspicious

File: tests/wallet/nft_wallet/test_nft_wallet.py [hunk 7]
```diff
@@ -332,7 +332,7 @@ async def ensure_wallet_sync() -> None:
     )
     assert len(txs) == 1
     assert txs[0].spend_bundle is not None
-    await wallet_node_1.wallet_state_manager.add_pending_transaction(txs[0])
```
Comment (arvidn) [inline tests/wallet/nft_wallet/test_nft_wallet.py:335]: here

File: tests/wallet/nft_wallet/test_nft_offers.py [hunk 8]
```diff
@@ -912,288 +905,3 @@ async def test_nft_offer_nft_for_nft(
 
     assert await nft_wallet_maker.get_nft_count() == 1
     assert await nft_wallet_taker.get_nft_count() == 1
-
-
-@pytest.mark.limit_consensus_modes(allowed=[ConsensusMode.PLAIN], reason="save time")
-@pytest.mark.parametrize("trusted", [True, False])
-@pytest.mark.parametrize("reuse_puzhash", [True, False])
-@pytest.mark.anyio
-async def test_nft_offer_nft0_and_xch_for_cat(
```
Comment (arvidn) [inline tests/wallet/nft_wallet/test_nft_offers.py:921]: why was this test removed?

File: tests/wallet/dao_wallet/test_dao_wallets.py [hunk 9]
```diff
@@ -225,7 +223,7 @@ async def test_dao_creation(self_hostname: str, two_wallet_nodes: OldSimulatorsA
 
     # send some cats from wallet_0 to wallet_1 so we can test voting
     cat_txs = await cat_wallet_0.generate_signed_transaction([cat_amt], [ph_1], DEFAULT_TX_CONFIG)
-    await wallet_0.wallet_state_manager.add_pending_transaction(cat_txs[0])
```
Comment (arvidn) [inline tests/wallet/dao_wallet/test_dao_wallets.py:228]: here

File: tests/wallet/dao_wallet/test_dao_wallets.py [hunk 10]
```diff
@@ -1014,7 +1004,7 @@ async def test_dao_proposal_partial_vote(
 
     cat_tx = await cat_wallet_0.generate_signed_transaction([100000], [ph_1], DEFAULT_TX_CONFIG)
     cat_sb = cat_tx[0].spend_bundle
-    await wallet_0.wallet_state_manager.add_pending_transaction(cat_tx[0])
```
Comment (arvidn) [inline tests/wallet/dao_wallet/test_dao_wallets.py:1017]: here

File: tests/wallet/dao_wallet/test_dao_wallets.py [hunk 11]
```diff
@@ -1121,7 +1109,7 @@ async def test_dao_proposal_partial_vote(
     ph_0 = await cat_wallet_0.get_new_inner_hash()
     cat_tx = await cat_wallet_1.generate_signed_transaction([balance + new_mint_amount], [ph_0], DEFAULT_TX_CONFIG)
     cat_sb = cat_tx[0].spend_bundle
-    await wallet_1.wallet_state_manager.add_pending_transaction(cat_tx[0])
```
Comment (arvidn) [inline tests/wallet/dao_wallet/test_dao_wallets.py:1124]: here

File: tests/wallet/dao_wallet/test_dao_wallets.py [hunk 12]
```diff
@@ -204,8 +203,7 @@ async def test_dao_creation(self_hostname: str, two_wallet_nodes: OldSimulatorsA
     # Send some cats to the dao_cat lockup
     dao_cat_amt = uint64(100)
     txs = await dao_wallet_0.enter_dao_cat_voting_mode(dao_cat_amt, DEFAULT_TX_CONFIG)
-    for tx in txs:
-        await wallet_0.wallet_state_manager.add_pending_transaction(tx)
```
Comment (arvidn) [inline tests/wallet/dao_wallet/test_dao_wallets.py:208]: there are a few places in here where you changed `wallet_0` to `dao_wallet_0`. are those referring to the same wallet or is there something subtle requiring these changes to the test?

File: tests/wallet/dao_wallet/test_dao_wallets.py [hunk 13]
```diff
@@ -3424,16 +3396,15 @@ async def test_dao_resync(self_hostname: str, two_wallet_nodes: OldSimulatorsAnd
...
+    await wallet_0.wallet_state_manager.add_pending_transactions(txs)
 
     await full_node_api.wait_transaction_records_entered_mempool(records=txs, timeout=60)
     await full_node_api.process_all_wallet_transactions(wallet_0, timeout=60)
     await full_node_api.wait_for_wallet_synced(wallet_node=wallet_node_0, timeout=30)
 
     # send some cats from wallet_0 to wallet_1 so we can test voting
     cat_txs = await cat_wallet_0.generate_signed_transaction([cat_amt], [ph_1], DEFAULT_TX_CONFIG)
-    await wallet_0.wallet_state_manager.add_pending_transaction(cat_txs[0])
```
Comment (arvidn) [inline tests/wallet/dao_wallet/test_dao_wallets.py:3436]: here

**Review by arvidn — COMMENTED** [review]
this looks like a really nice improvement!
I had a few comments of things that look a bit suspicious

**Review by matt-o-how — DISMISSED** [review]
Looks good to me

File: chia/pools/pool_wallet.py (hunk 1 above)
Comment (Quexington) [reply chia/pools/pool_wallet.py:892]: Yeah it was a bit of a decision I made because of a sort of bugged (but working way) it was operating before:
The way it used to work:
* fee tx would have its own spend bundle
* travel tx would get all spends attached to it including the fee spend
* travel tx would be pushed first
* the fee tx is optionally pushed

This works because travel tx gets to the mempool first and then the fee tx is just a mempool duplicate that gets rejected as such.  However, when we start pushing both transactions with the new method, we are sure to aggregate the spends from all transactions together onto a single record to make sure they are pushed at once.  This means that we had two copies of the fee spend in the spendbundle after aggregation because we had included it on both txs.

Instead of choosing not to include the fee spend on the travel tx, I just adopted the existing paradigm that most wallets use and do the (now unnecessary) aggregation and fee tx clearance before we send the transactions to the DB to get pushed.

File: chia/wallet/wallet_state_manager.py (hunk 2 above)
Comment (Quexington) [reply chia/wallet/wallet_state_manager.py:2232]: No, as far as I know we have no commitment to the uniformity of transaction IDs so the `name` property is mostly used as a de-dup and random 32 bytes works great for that.

Comment (Quexington) [reply chia/wallet/wallet_state_manager.py:2232]: Worth noting that I considered just using the existing name here but some wallets set that name property to the spend bundle name (which this function is already doing) and we end up with duplicates.  Rather than have magic knowledge to not accidentally do, I went this route.

File: tests/wallet/dao_wallet/test_dao_wallets.py (hunk 12 above)
Comment (Quexington) [reply tests/wallet/dao_wallet/test_dao_wallets.py:208]: They don't refer to the same wallet but their `.wallet_state_manager` property is the same.  The reason I changed it is because it is important you don't actually use a wallet state manager from a different simulated environment and using the same wallet as the one that generated the transactions seemed like safer in that regard.

Comment (Quexington) [reply tests/wallet/dao_wallet/test_dao_wallets.py:208]: (I'm going to resolve the other comments where you marked this as occuring)

File: tests/wallet/nft_wallet/test_nft_offers.py (hunk 8 above)
Comment (Quexington) [reply tests/wallet/nft_wallet/test_nft_offers.py:921]: Good catch, the answer is "on accident" :)

File: chia/pools/pool_wallet.py (hunk 1 above)
Comment (arvidn) [reply chia/pools/pool_wallet.py:892]: it might warrant a comment in the code. but I'll leave that up to you. feel free to resolve this thread

**Review by arvidn — APPROVED** [review]
very nice!
